_posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd

```diff
@@ -0,0 +1,492 @@
+  (3): R2Conv([C4_on_R2[(None, 4)]:
+       {regular (x4)}(16)], [C4_on_R2[(None, 4)]: {regular (x4)}(16)], kernel_size=3, stride=1)
+  (4): InnerBatchNorm([C4_on_R2[(None, 4)]:
+       {regular (x4)}(16)], eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
+  (5): ReLU(inplace=False, type=[C4_on_R2[(None, 4)]: {regular (x4)}(16)])
+  (6): R2Conv([C4_on_R2[(None, 4)]:
+       {regular (x4)}(16)], [C4_on_R2[(None, 4)]: {regular (x1)}(4)], kernel_size=3, stride=1)
+)
+```
+
+Calling this model on some input image, we get:
+
+```{r}
+x <- torch$randn(1L, 1L, 17L, 17L)
+x <- feat_type_in(x)
+model(x)$shape |> unlist()
+```
+
+```         
+[1]  1  4 11 11
+```
+
+What we do now depends on the task. Since we didn't preserve the original resolution anyway -- as would have been required for, say, segmentation -- we probably want one feature vector per image. That we can achieve by spatial pooling:
+
+```{r}
+avgpool <- nn$PointwiseAvgPool(feat_type_out, 11L)
+y <- avgpool(model(x))
+y$shape |> unlist()
+```
+
+```         
+[1] 1 4 1 1
+```
+
+We still have four "channels", corresponding to four group elements. This feature vector is (approximately) translation-*in*variant, but rotation-*equi*variant, in the sense expressed by the choice of group. Often, the final output will be expected to be group-invariant as well as translation-invariant (as in image classification). If that's the case, we pool over group elements, as well:
+
+```{r}
+invariant_map <- nn$GroupPooling(feat_type_out)
+y <- invariant_map(avgpool(model(x)))
+y$tensor
+```
+
+```         
+tensor([[[[-0.0293]]]], grad_fn=<CopySlices>)
+```
+
+We end up with an architecture that, from the outside, will look like a standard convnet, while on the inside, all convolutions have been performed in a rotation-equivariant way. Training and evaluation then are no different from the usual procedure.
+
+## Where to from here
+
+This "introduction to an introduction" has been the attempt to draw a high-level map of the terrain, so you can decide if this is useful to you. If it's not just useful, but interesting theory-wise, as well, you'll find lots of excellent materials linked from the [README](https://github.com/QUVA-Lab/escnn). The way I see it, though, this post already should enable you to actually experiment with different setups.
+
+One such experiment, that would be of high interest to me, might investigate how well different types and degrees of equivariance actually work for a given task and dataset. Overall, a reasonable assumption is that, the "up" we go in the feature hierarchy, the less equivariance we require. For edges and corners, taken by themselves, full rotation equivariance seems desirable, as does equivariance to reflection; for higher-level features, we might want to successively restrict allowed operations, maybe ending up with equivariance to mirroring merely. Experiments could be designed to compare different ways, and levels, of restriction.[^7]
```

Comment:
```suggestion
One such experiment, that would be of high interest to me, might investigate how well different types and degrees of equivariance actually work for a given task and dataset. Overall, a reasonable assumption is that, the higher "up" we go in the feature hierarchy, the less equivariance we require. For edges and corners, taken by themselves, full rotation equivariance seems desirable, as does equivariance to reflection; for higher-level features, we might want to successively restrict allowed operations, maybe ending up with equivariance to mirroring merely. Experiments could be designed to compare different ways, and levels, of restriction.[^7]
```

Reply:
Or, if you don't want to use higher/lower, you can use more/less, increase/decreate, etc.—it's just that "the up we go…" doesn't read correctly.